Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+  , a ~ PAsData a'
+  , PIsData a'
+  , PTryFrom PData b
+  , b ~ PAsData b'
+  , PIsData b'
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinPair a b))
+  where
+  type PExcess PData (PAsData (PBuiltinPair a b)) = PBuiltinPair a b
+  ptryFrom opq = do
+    tup <- tcont $ plet (pfromData $ punsafeCoerce opq)
+    let fst' :: Term _ a
+        fst' = unTermCont $ fst <$> (ptryFrom @PData @a $ pforgetData $ pfstBuiltin # tup)
+        snd' :: Term _ b
+        snd' = unTermCont $ fst <$> (ptryFrom @PData @b $ pforgetData $ psndBuiltin # tup)
+    ver <- tcont $ plet $ ppairDataBuiltin # fst' # snd'
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData (PAsData b)
+  , PTryFrom PData (PAsData (PDataRecord xs))
+  , x ~ (s ':= b)
+  , PIsData (PDataRecord xs)
+  , PExcess PData (PAsData (PDataRecord xs)) ~ PDataRecord xs
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord (x ': xs)))
+  where
+  type PExcess PData (PAsData (PDataRecord (x ': xs))) = PDataRecord (x ': xs)
+  ptryFrom opq = do
+    lst <- tcont $ plet (pfromData @(PBuiltinList _) $ punsafeCoerce opq)
+    let lsthead :: Term _ PData
+        lsthead = phead # lst
+        lsttail :: Term _ (PAsData (PBuiltinList PData))
+        lsttail = pdata $ ptail # lst
+        verhead :: Term _ (PAsData b)
+        verhead = unTermCont $ fst <$> (ptryFrom @PData @(PAsData b) lsthead)
+        vertail :: Term _ (PDataRecord xs)
+        vertail = unTermCont $ snd <$> (ptryFrom @PData @(PAsData (PDataRecord xs)) (pforgetData lsttail))
+    ver <-
+      tcont $
+        plet $
+          pdcons @s
+            # verhead
+            # vertail
+    pure (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPING #-}
+  ( PTryFrom PData (PAsData b)
+  , PIsData b
+  , x ~ (s ':= b)
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord '[x]))
+  where
+  type PExcess PData (PAsData (PDataRecord '[x])) = PDataRecord '[x]
+  ptryFrom opq = do
+    let lsthead :: Term _ PData
+        lsthead = phead # (pfromData @(PBuiltinList _) $ punsafeCoerce opq)
+        verhead :: Term _ (PAsData b)
+        verhead = unTermCont $ fst <$> (ptryFrom @PData @(PAsData b) lsthead)
+    ver <-
+      tcont $
+        plet $
+          pdcons @s
+            # verhead
+            # pdnil
+    pure $ (punsafeCoerce opq, ver)
+
+----------------------- POpaque Instances -----------------------------------------------
+
+{- |
+    for none of the opaque instances it can be verified
+    that the actual structure is what it says to be
+    because that data is lost when the PAsData wrapper
+    is removed, this can only be safely used if you obtained
+    your POpaque safely
+-}
+instance
+  ( PTryFrom PData (PAsData a)
+  , PIsData a
+  ) =>
+  PTryFrom POpaque a
+  where
+  type PExcess POpaque a = PAsData a
+  ptryFrom opq = do
+    let prop :: Term _ a
+        prop = punsafeCoerce opq
+    ver' <- fst <$> (ptryFrom @PData @(PAsData a) $ pforgetData $ pdata prop)
+    ver <- tcont $ plet ver'
```

Comment:
NB: I don't think you need the `plet` here technically, albeit it should be optimised out anyway, since `plet x` where `x` is an `RVar _` inlines the binding on use sites. Because of this you should just let this be as it is now.